src/main/java/de/tum/in/ase/notification/NotificationPlugin.java

```diff
@@ -0,0 +1,174 @@
+package de.tum.in.ase.notification;
+
+import com.google.gson.Gson;
+import de.tum.in.ase.notification.configuration.Context;
+import de.tum.in.ase.notification.configuration.ContextFactory;
+import de.tum.in.ase.notification.exception.TestParsingException;
+import de.tum.in.ase.notification.model.Commit;
+import de.tum.in.ase.notification.model.TestResults;
+import de.tum.in.ase.notification.model.Testsuite;
+import org.apache.commons.io.IOUtils;
+import org.apache.http.HttpException;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.fluent.Request;
+import org.apache.http.entity.ContentType;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import javax.xml.bind.JAXBContext;
+import javax.xml.bind.JAXBException;
+import javax.xml.bind.Unmarshaller;
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * Abstract class for the notification plugin. All other subtypes of the plugin, e.g. the CLI plugin, should extend this as this class provides the basic functionality.
+ */
+public abstract class NotificationPlugin {
+    private static final Logger LOGGER = LogManager.getLogger();
+    protected final ContextFactory contextFactory;
+
+    protected NotificationPlugin(ContextFactory contextFactory) {
+        this.contextFactory = contextFactory;
+    }
+
+    /**
+     * Runs the plugin. This includes reading the test results, parsing them and sending them to the Artemis server.
+     *
+     * @param context the context containing the configuration.
+     * @throws IOException if one of the result directories is not readable.
+     */
+    public void run(Context context) throws IOException {
+        final Path currentPath = Paths.get(".").toAbsolutePath().normalize();
+        final Path testResultsDir = currentPath.resolve(Paths.get(context.getTestResultsDir()));
+        final Path customFeedbackDir = currentPath.resolve(Paths.get(context.getCustomFeedbackDir()));
+        final Path buildLogsFile = currentPath.resolve(Paths.get(context.getBuildLogsFile()));
+
+        final List<Testsuite> testReports = extractTestResults(testResultsDir);
+        final Optional<Testsuite> customFeedback = CustomFeedbackParser.extractCustomFeedbacks(customFeedbackDir);
+        customFeedback.ifPresent(testReports::add);
+
+        final TestResults results = combineTestResults(testReports, context);
+
+        results.setIsBuildSuccessful(context.getBuildStatus().equalsIgnoreCase("success"));
```

Comment:
Can the `buildStatus` be `null`? Maybe a small method in the context `boolean isBuildSuccessful()` that also checks for `null` first would be helpful.

Edit: The `EnvReader` should ensure non-null. I still think encapsulating the logic in a helper method makes sense.